File: PHTV/UI/Models/KlipyModels.swift
//
//  KlipyModels.swift
//  PHTV
//
//  Created by Phạm Hùng Tiến on 2026.
//  Copyright © 2026 Phạm Hùng Tiến. All rights reserved.
//

import Foundation

// MARK: - Klipy API Response Models

struct KlipyResponse: Codable {
    let result: Bool
    let data: KlipyData
}

struct KlipyData: Codable {
    let data: [KlipyGIF]
    let current_page: Int
    let per_page: Int
    let has_next: Bool
}

struct KlipyGIF: Codable, Identifiable {
    let id: Int64
    let slug: String
    let title: String
    let file: KlipyFile
    let tags: [String]?  // Optional - API có thể trả về null
    let type: String

    // Ad-specific fields (optional, only present when type == "ad")
    let impression_url: String?
    let click_url: String?
    let target_url: String?
    let advertiser: String?

    var isAd: Bool {
        type == "ad"
    }

    var previewURL: String {
        // Use small size for preview, fallback to any available size
        file.sm?.gif?.url ?? file.xs?.gif?.url ?? file.hd?.gif?.url ?? ""
    }

    var fullURL: String {
        // Use HD GIF for full quality, fallback to smaller sizes
        file.hd?.gif?.url ?? file.sm?.gif?.url ?? file.xs?.gif?.url ?? ""
    }

    // Safe check for empty URLs
    var hasValidURL: Bool {
        !previewURL.isEmpty && !fullURL.isEmpty
    }
}

struct KlipyFile: Codable {
    let hd: KlipyFileSize?  // Optional - API có thể không trả về hd
    let sm: KlipyFileSize?
    let xs: KlipyFileSize?
}

struct KlipyFileSize: Codable {
    let gif: KlipyMedia?
    let webp: KlipyMedia?
    let mp4: KlipyMedia?
}

struct KlipyMedia: Codable {
    let url: String
    let width: Int?
    let height: Int?
    let size: Int?
}

File: Windows/App/Models/KlipyModels.cs
using System.Text.Json.Serialization;

namespace PHTV.Windows.Models;

public sealed class KlipyResponse {
    [JsonPropertyName("result")]
    public bool Result { get; set; }

    [JsonPropertyName("data")]
    public KlipyData Data { get; set; } = new();
}

public sealed class KlipyData {
    [JsonPropertyName("data")]
    public List<KlipyGif> Data { get; set; } = new();

    [JsonPropertyName("current_page")]
    public int CurrentPage { get; set; }

    [JsonPropertyName("per_page")]
    public int PerPage { get; set; }

    [JsonPropertyName("has_next")]
    public bool HasNext { get; set; }
}

public sealed class KlipyGif {
    [JsonPropertyName("id")]
    public long Id { get; set; }

    [JsonPropertyName("slug")]
    public string Slug { get; set; } = string.Empty;

    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;

    [JsonPropertyName("file")]
    public KlipyFile File { get; set; } = new();

    [JsonPropertyName("tags")]
    public List<string>? Tags { get; set; }

    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("impression_url")]
    public string? ImpressionUrl { get; set; }

    [JsonPropertyName("click_url")]
    public string? ClickUrl { get; set; }

    [JsonPropertyName("target_url")]
    public string? TargetUrl { get; set; }

    [JsonPropertyName("advertiser")]
    public string? Advertiser { get; set; }

    [JsonIgnore]
    public bool IsAd => string.Equals(Type, "ad", StringComparison.OrdinalIgnoreCase);

    [JsonIgnore]
    public string PreviewUrl =>
        File.Sm?.Gif?.Url ?? File.Xs?.Gif?.Url ?? File.Hd?.Gif?.Url ?? string.Empty;

    [JsonIgnore]
    public string FullUrl =>
        File.Hd?.Gif?.Url ?? File.Sm?.Gif?.Url ?? File.Xs?.Gif?.Url ?? string.Empty;

    [JsonIgnore]
    public bool HasValidUrl => !string.IsNullOrEmpty(PreviewUrl) && !string.IsNullOrEmpty(FullUrl);
}

public sealed class KlipyFile {
    [JsonPropertyName("hd")]
    public KlipyFileSize? Hd { get; set; }

    [JsonPropertyName("sm")]
    public KlipyFileSize? Sm { get; set; }

    [JsonPropertyName("xs")]
    public KlipyFileSize? Xs { get; set; }
}

public sealed class KlipyFileSize {
    [JsonPropertyName("gif")]
    public KlipyMedia? Gif { get; set; }

    [JsonPropertyName("webp")]
    public KlipyMedia? Webp { get; set; }

    [JsonPropertyName("mp4")]
    public KlipyMedia? Mp4 { get; set; }
}

public sealed class KlipyMedia {
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("width")]
    public int? Width { get; set; }

    [JsonPropertyName("height")]
    public int? Height { get; set; }

    [JsonPropertyName("size")]
    public int? Size { get; set; }
}
